### PR TITLE
Implemented product buy action

### LINKED ISSUE #254: Product reservation
There is currently no back-end to reserve/"buy" products.

### PR DESCRIPTION
**Pull request by Naapperas**
Closes #254

### CODE REVIEW

**Review by toni-santos — COMMENTED** [review]
Looks good and seems to work, however we should make users aware that products that have been bought cannot be bought again. This shouldn't be too hard and could be done in this PR.

File: app/Http/Controllers/ShopController.php [hunk 1]
```diff
@@ -29,4 +32,42 @@ public function show(Request $request)
...
+
+        if ($currentEnrollment === null) {
+            return redirect()->back()->dangerBanner('You are not enrolled in this edition');
+        }
+
+        $points = $currentEnrollment->points;
+
+        if ($points < $product->price) {
+            return redirect()->back()->dangerBanner('You do not have enough points');
+        }
```
Comment (ttoino) [inline app/Http/Controllers/ShopController.php]: This should be done in a gate or policy, and the error message(s) should be in portuguese

File: app/Http/Controllers/ShopController.php [hunk 2]
```diff
@@ -29,4 +32,42 @@ public function show(Request $request)
...
+
+        $currentEnrollment->decrement('points', $product->price);
+        $currentEnrollment->products()->attach($product->id);
+
+        return redirect()->back()->banner('Product bought');
```
Comment (ttoino) [inline app/Http/Controllers/ShopController.php]: This should be in portuguese

File: app/Http/Controllers/ShopController.php [hunk 3]
```diff
@@ -29,4 +32,42 @@ public function show(Request $request)
...
+            return redirect()->back()->dangerBanner('You are not enrolled in this edition');
+        }
+
+        $points = $currentEnrollment->points;
+
+        if ($points < $product->price) {
+            return redirect()->back()->dangerBanner('You do not have enough points');
+        }
+
+        $currentEnrollment->decrement('points', $product->price);
```
Comment (ttoino) [inline app/Http/Controllers/ShopController.php]: Don't we have a trigger for this?

File: resources/js/Components/Shop/ShopItem.vue [hunk 4]
```diff
@@ -58,8 +65,7 @@ defineProps<Props>();
                 Confirmar compra de <b>{{ product.name }}</b> por
                 {{ product.price }}?
             </p>
-            <!-- TODO: BACKEND CONNECTION -->
-            <PrimaryButton>Comprar</PrimaryButton>
+            <PrimaryButton @click="buyProduct()">Comprar</PrimaryButton>
```
Comment (ttoino) [inline resources/js/Components/Shop/ShopItem.vue:70]: We should check the aforementioned gate here, so the user doesn't see the button if they can't buy the product.

**Review by ttoino — CHANGES_REQUESTED** [review]

File: app/Http/Controllers/ShopController.php (hunk 3 above)
Comment (toni-santos) [reply app/Http/Controllers/ShopController.php]: Also, isn't in portuguese

Comment (Naapperas) [reply app/Http/Controllers/ShopController.php]: We do not @ttoino, should I create it?

Comment (Naapperas) [reply app/Http/Controllers/ShopController.php]: Will create it.

File: resources/js/Components/Shop/ShopItem.vue (hunk 4 above)
Comment (Naapperas) [reply resources/js/Components/Shop/ShopItem.vue:70]: :point_up: :nerd_face: "aforementioned"

File: app/Http/Controllers/ShopController.php [hunk 5]
```diff
@@ -15,18 +17,48 @@ public function show(Request $request)
         if ($user && $user->isParticipant()) {
             $isParticipant = $user->isParticipant();
             $isEnrolled = $user->usertype->enrollments()->where('edition_id', $edition->id)->exists();
-            $points = $user->usertype->enrollments()->where('edition_id', $edition->id)->first()->points;
+
+            $points = $isEnrolled ? $user->usertype->enrollments()->where('edition_id', $edition->id)->first()->points : null;
```
Comment (ttoino) [inline app/Http/Controllers/ShopController.php]: This queries the database for the enrollment twice

**Review by ttoino — DISMISSED** [review]

**Review by toni-santos — APPROVED** [review]

**Pull request merged**